### PR TITLE
Views integration specs n+1 problems

### PR DESCRIPTION
**Pull request by chudisoft**
Project requirements
n+1

    Make sure that the N+1 problem is solved when fetching all posts and their comments for a user.
        Do not use the bullet gem in this exercise - instead analyze the logs as showed in this [article](https://dev.to/junko911/rails-n-1-queries-and-eager-loading-10eh)
        In your PR description include the screenshots of your app's logs before and after the fix
        This is the corresponding wireframe view: blog_user_all_posts

Integration specs

    Use Capybara to write integration tests for each view in your project.

    Think on the flow of things and group the tests of page in a way that makes sense. For example, one test would check for the things in the page and another test for clicking links on it.

    User index page:
        I can see the username of all other users.
        I can see the profile picture for each user.
        I can see the number of posts each user has written.
        When I click on a user, I am redirected to that user's show page.

    user show page:
        I can see the user's profile picture.
        I can see the user's username.
        I can see the number of posts the user has written.
        I can see the user's bio.
        I can see the user's first 3 posts.
        I can see a button that lets me view all of a user's posts.
        When I click a user's post, it redirects me to that post's show page.
        When I click to see all posts, it redirects me to the user's post's index page.

    User post index page:
        I can see the user's profile picture.
        I can see the user's username.
        I can see the number of posts the user has written.
        I can see a post's title.
        I can see some of the post's body.
        I can see the first comments on a post.
        I can see how many comments a post has.
        I can see how many likes a post has.
        I can see a section for pagination if there are more posts than fit on the view.
        When I click on a post, it redirects me to that post's show page.

    Post show page:
        I can see the post's title.
        I can see who wrote the post.
        I can see how many comments it has.
        I can see how many likes it has.
        I can see the post body.
        I can see the username of each commentor.
        I can see the comment each commentor left.
        
        Our tests results below 👇🏾👇🏾👇🏾
image

These are the queries before solving N+1 problem for fetching all posts with 5 recent comments: 👇🏾👇🏾👇🏾

image
image

These are the queries after solving N+1 problem for fetching all comments for a selected post: 👇🏾👇🏾👇🏾

image

### CODE REVIEW

File: spec/spec/features/user_posts_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        unique_comment_content = "#{comment.user.name}: #{comment.text}"
+        expect(page).to have_content(unique_comment_content)
+      end
+    end
+  end
+
+  it 'redirects to a post show page when a post is clicked' do
+    post = user.posts.first
+    click_on post.title
+    expect(current_path).to eq(user_post_path(user, post))
```
Comment (Ibizugbe) [inline spec/spec/features/user_posts_index_spec.rb]: Nice job so far👍🏼. As a[ requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md), kindly add a Test that asserts you can see a section for pagination if there are more posts than fit on the view👍🏼.

image

File: spec/spec/features/user_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+  it 'shows the number of posts each user has written' do
+    User.all.each do |user|
+      expect(page).to have_content("Number of posts: #{user.posts.count}")
+    end
+  end
+
+  it 'redirects to user show page on clicking a user' do
+    user = User.first
+    # click_on user.name
```
Comment (Ibizugbe) [inline spec/spec/features/user_index_spec.rb]: Kindly do well to take every commented piece of code from your spec files as it gives your codebase a messy look😉.

**Review by Ibizugbe — CHANGES_REQUESTED** [review]
Hi @chudisoft  @GhostEsso 👋🏼👋🏼👋🏼, 


request change


Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights✅✅✅

- [x] All tests are passing👍🏼.
- [x] Tests cases for the user_index, user_show have been captured👍🏼.


## Required Changes ♻️

- [ ] As a [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md), kindly add the screenshots of your app's logs before and after the fix😉.
image



_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/spec/features/user_index_spec.rb (hunk 2 above)
Comment (LionRouge1) [reply spec/spec/features/user_index_spec.rb]: - [ ] As mentioned by the previous reviewer, Kindly remove all commented out code from your code base. It make your code looks messy and unclean

File: spec/spec/features/user_posts_index_spec.rb (hunk 1 above)
Comment (LionRouge1) [reply spec/spec/features/user_posts_index_spec.rb]: - [ ] As pointed out by the previous reviewer, Kindly add test that check if the pagination button is present when there are more posts to follow the project requirement.

**Review by LionRouge1 — CHANGES_REQUESTED** [review]
Hi @chudisoft, 

Great work on making the changes 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

almost

## Suggested changes :recycle: 

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mckent05 — APPROVED** [review]
Hi @chudisoft, 

# STATUS: APPROVED ✔️ ✔️ 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations and well done! 🎉 
approved


## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👏 👏 
- Good job solving the N+1 queries and ensuring all tests pass 👍 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**